python_modules/dagster/dagster/_config/pythonic_config/__init__.py

```diff
@@ -812,16 +825,56 @@ def with_resource_context(
         return copy
 
     def _initialize_and_run(self, context: InitResourceContext) -> TResValue:
-        updated_resource = (
-            self._resolve_and_update_nested_resources(context)  # noqa: SLF001
-            .with_resource_context(context)
-            ._resolve_and_update_env_vars()
-        )
+        with self._resolve_and_update_nested_resources(context) as has_nested_resource:
+            updated_resource = has_nested_resource.with_resource_context(  # noqa: SLF001
+                context
+            )._resolve_and_update_env_vars()
 
-        return updated_resource._create_object_fn(context)  # noqa: SLF001
+            updated_resource.pre_execute(context)
+            return updated_resource.create_resource(context)
 
-    def _create_object_fn(self, context: InitResourceContext) -> TResValue:
-        return self.create_resource(context)
+    @contextlib.contextmanager
+    def _initialize_and_run_cm(
+        self, context: InitResourceContext
+    ) -> Generator[TResValue, None, None]:
+        with self._resolve_and_update_nested_resources(context) as has_nested_resource:
+            updated_resource = has_nested_resource.with_resource_context(  # noqa: SLF001
+                context
+            )._resolve_and_update_env_vars()
+
+            with updated_resource.yield_for_execution(context) as value:
+                yield value
```

Comment:
Having these dual code paths is confusing. Seems like `_initialize_and_run_cm` should be the canonical implementation here